docs/intro-to-coalton-testing.md

```diff
@@ -0,0 +1,220 @@
+
+The `is` macro asserts that its first argument evaluates to true. `(is True)` will always
+succeed, and `(is False)` will always fail.
+
+If you have a function `always-returns-zero` in your package `foo`, and you want to make
+sure it returns `0`, you might write:
+
+```
+(define-test test-always-returns-zero ()
+  (is (== 0 (foo:always-returns-zero))))
+```
+
+When calling boolean predicates in an `is` form, like `==`, prefer writing the expected
+value as the first operand, and the computed value as the second.
+
+TODO: `coalton-testing` doesn't actually privilege that argument order the way FiveAM and
+      (I assume) Fiasco do. Should we recommend it anyway? I still think it's a nice
+      style, personally.
+
+`is` optionally takes a second argument, a string used to describe the assertion if it
+fails. For example, you might write:
+
+```
+(is (== 0 (foo:always-returns-zero))
+    "ALWAYS-RETURNS-ZERO returned a non-zero value!")
+```
+
+### Pattern assertions with `matches`
+
+The `matches` macro takes two required arguments, a pattern and an expression, and asserts
+that the result of the expression matches the pattern. `(matches _ ANYTHING)` will always
+succeed (assuming `anything` compiles and executes without error).
+
+If you have a function `one-element-list` which wraps its argument in a one-element list
+in your package `foo`, you might write:
+
+```
+(define-test test-one-element-list ()
+  (matches (Cons _ (Nil))
+           (foo:one-element-list 0)))
+```
+
+Note that this doesn't assert anything about the contents of the list returned my
+`(foo:one-element-list 0)`, only its structure. If `(foo:one-element-list 0)` returns the
+list `(Cons 0 (Nil))`, the test will pass just the same as if it returned `(Cons "zero"
+(Nil))`.
+
+Like `is`, `matches` takes an optional string argument used to describe the assertion if
+it fails. You might write:
+
+```
+(matches (Cons _ (Nil))
+         (foo:one-element-list 0)
+         "ONE-ELEMENT-LIST returned a list with a length other than 1!")
+```
+
+### Uses for `(is False)`
+
+Often, it's useful to invoke a function, destructure its return with `match`, and to run
+additional assertions in some branches while treating the other branches as failures. In
+this case, it's convenient to write an `(is False)` assertion with a message describing
+the branch incorrectly taken in each of the failed branches. For example, if our
+`one-element-list` function from above is supposed to wrap specifically its argument in a
+list, we'll want to add an `==` assertion. We might write:
+
+```
+(define-test test-one-element-list ()
+  (match (one-element-list 0)
+    ((Nil) (is False "ONE-ELEMENT-LIST returned an empty list!"))
+    ((Cons elt (Nil)) (is (== 0 elt)))
+    ((Cons _ _) (is False "ONE-ELEMENT-LIST returned a list with more than 1 element!"))))
+```
+
+## Running your tests
+
+Now that you've defined a test system and some tests to go in it, you can run your tests
```

Comment:
There should be a note here that running asdf commands should not be done in a coalton package. Doing so tends to corrupt asdf.

Reply:
Is this still the case now that `coalton` reexports `defpackage` and `in-package`? And if so, is it a breakage between Coalton and ASDF, or is it that some other systems expect to be in a package that uses CL before being loaded, and contain code with that expectation before defining and entering their own packages?